encoding/codecs/h264.md

```diff
@@ -0,0 +1,247 @@
+In this case, we want to chain `vspipe`,
+the program that reads VapourSynth scripts,
+with x264, our encoder.
+
+#### `--demuxer y4m`
+
+This tells x264 that we're providing it with a y4m file.
+This matches up with the `--y4m` flag that we gave to the `vspipe` command.
+
+#### `--preset veryfast`
+
+x264 has a set of presets to switch between faster encoding, or higher quality.
+The full list of presets, from fastest to slowest, is:
+
+- ultrafast
+- superfast
+- veryfast
+- faster
+- fast
+- medium
+- slow
+- slower
+- veryslow
+- placebo
+
+You will almost never want to use the very extreme settings,
+but generally, if you want good quality
+and don't care about how long the encode takes,
+`slower` or `veryslow` are recommended.
+In this example, because we are just demonstrating how x264 works,
+we want a fast encode and have chosen `veryfast`.
+
+For the curious, you can see a full list of the settings
+enabled by each preset by running `x264 --fullhelp | less`.
+However, this probably won't mean much at the moment.
+Don't worry, this guide will explain later what all of those settings mean.
+
+#### `--tune animation`
+
+Beyond the preset chosen,
+x264 allows us to further tune the encoding settings
+for the type of content we're working with.
+The following tunings are generally the most useful:
+
+- `film`: Recommended for live action videos.
+- `animation`: Recommended for anime or cartoons with flat textures.
+For 3D animation (e.g. modern Pixar movies),
+you may find better results with `film`.
+- `grain`: Recommended for particularly grainy films.
+
+You don't need to use a tuning,
+but it generally helps to produce a better-looking video.
+
+#### `--crf 24`
+
+CRF is a constant-quality, 1-pass encoding mode.
+In layman's terms,
+this means that we don't need the output to meet a specific filesize,
+we just want the output to meet a certain quality level.
+CRF ranges from 0 to 51,
+with 0 being the best quality and 51 being the smallest filesize,
+but there is a certain range of CRF settings
+that are generally most useful.
+Here are some guidelines:
+
+- CRF 13: This is considered visually lossless to videophiles.
+This can produce rather large files,
+but is a good choice if you want high quality videos.
+Fansubbing groups often use this for Blu-ray encodes.
+- CRF 16-18: This is considered visually lossless to most viewers,
+and leans toward high quality while still providing a reasonable filesize.
+This is a typical range for fansub encodes.
+- CRF 21-24: This provides a good balance between quality and filesize.
+Some quality loss becomes visible, but this is generally a good choice
+where filesize becomes a concern, such as for videos viewed over the internet.
+- CRF 26-30: This prioritizes filesize, and quality loss becomes more obvious.
+It is generally not recommended to go higher than CRF 30 in any real-world
+encoding scenario, unless you want your videos to look like they were made
+for dial-up.
```

Comment:
What feels odd about this section is that the ranges aren’t continuous. What about the range between 13 and 16? 19 and 20? Bluray encodes often are in the range 13-15, so you could just extend these ranges a little to not leave any gaps.

Reply:
I considered this too, but decided the reader should be smart enough to recognize that anything outside the regions specified is either, well, not in any of these regions or only for special occasions. It's hard enough to define definite regions for CRF as is already, so leaving a few gaps leaves more room for interpretation.

Now, what we should mention is that CRF is slightly dependant on resolution, or rather what videophiles generally consider as transparent or acceptable.